### PR TITLE
Add `jsx-import` rule

### LINKED ISSUE #155: Rule proposal: `jsx-import`
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/155)

Idea from: https://github.com/sindresorhus/refined-github/pull/1089/files#r168391471

When JSX is detected in the source, require the code to import one of the following imports:

- `import React from 'react';`
- `import preact from 'preact';`
- `import {h} from 'preact';`
- `import {h} from 'dom-chef';`
- … (We can fill this list out from need)

And if no JSX is is detect, disallow these imports.

Need suggestions for a better rule name.

Thoughts on the rule?

> There is a $40.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/155).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by doniyor2109**
Fixes #155.
Continue of PR #162

### CODE REVIEW

**Comment by sindresorhus** [comment]
Hey, thanks for the pull request 🙌 I have gotten a huge response on the IssueHunt bounty issues and I now have a very long backlog of pull requests to review, so it might take some time until I will be able to review this. However, you can help move this along by helping out and reviewing the [other submissions](https://github.com/sindresorhus/eslint-plugin-unicorn/pulls). If everyone helps everyone we can manage this 👌.

Some things to look for when reviewing:

- Ensure the rule addition adheres to the [instructions](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/new-rule.md).
- There are never too many tests. Try to find both passing and failing cases, especially obscure cases.
- Check the docs for typos and whether something could be more clear.
- Look at the rule code and see if anything could be simplified or done better. If something is unclear in the code, ask the submitter to explain.

**Comment by doniyor2109** [comment]
@sindresorhus Yea, after IssueHunt there are more PRs. It is pleasure to help with reviewing PRs.

I have automated checking docs and tests for new rule in #234

File: docs/rules/jsx-import.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Enforce importing a pragma when using JSX
+
+This rule enforces importing a pragma when JSX is used in the file. This is so that the JSX can be transformed into JavaScript. For example, when using `React`, `<div/>` is transformed to `React.createElement('div')`.
+
+
```
Comment (sindresorhus) [inline docs/rules/jsx-import.md]: The documentation should list what pragmas the rule supports.

File: test/jsx-import.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+		'import preact from \'preact\'; let Foo; <Foo/>;',
+		'import preact, {render} from \'preact\'; let Foo; <Foo/>;',
+		'import {h, render} from \'preact\'; let Foo; <Foo/>;',
+		'import {h} from \'dom-chef\'; let Foo; <Foo/>;',
+		'import React from \'react\'; import bar from \'bar\'; let Foo; <Foo/>;',
+		'const React = require(\'react\'); let Foo; <Foo/>;',
+		'const preact = require(\'preact\'); let Foo; <Foo/>;',
+		'const {h} = require(\'preact\'); let Foo; <Foo/>;',
+		'const {h} = require(\'dom-chef\'); let Foo; <Foo/>;',
+		'const {h} = require(\'dom-chef\'); const bar = require(\'bar\'); let Foo; <Foo/>;',
```
Comment (sindresorhus) [inline test/jsx-import.js]: `dom-chef` supports using `import React from \'dom-chef\';` now, so that should be tested.

File: rules/jsx-import.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	React: ['cloneElement']
+};
+
+const getSuggestedPackages = modules => arrayToSentence(modules.map(x => `\`${x}\``), {lastSeparator: ' or '});
+
+const hasPackage = (node, packages) => node.init ?
+	packages.some(x => x === astUtils.getRequireSource(node.init)) :
+	packages.some(x => x === node.parent.source.value);
+
+const isImportDeclaration = node => node.init ?
```
Comment (sindresorhus) [inline rules/jsx-import.js]: This could be moved to a `index.js` file in https://github.com/sindresorhus/eslint-plugin-unicorn/tree/master/rules/utils

File: rules/jsx-import.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	astUtils.isStaticRequire(node.init) :
+	node.parent.type === 'ImportDeclaration';
+
+const getVariable = (name, variables) => variables.find(x => x.name === name);
+
+// https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/util/variable.js#L27-L52
+const getVariablesInScope = context => {
```
Comment (sindresorhus) [inline rules/jsx-import.js]: This could be moved to a `get-variables-in-scope.js` file in https://github.com/sindresorhus/eslint-plugin-unicorn/tree/master/rules/utils

File: test/jsx-import.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+		},
+		{
+			code: 'const React = require(\'foobar\'); let Foo; <Foo/>',
+			errors: [wrongError('React')],
+			output: 'const React = require(\'react\'); let Foo; <Foo/>'
+		},
+		{
+			code: 'const preact = require(\'foobar\'); let Foo; <Foo/>',
+			errors: [wrongError('preact')],
+			output: 'const preact = require(\'preact\'); let Foo; <Foo/>'
```
Comment (sindresorhus) [inline test/jsx-import.js]: I don't think we should auto-fix the module name. This is best left as just an error.

File: rules/jsx-import.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+				isJsx = false;
+				isModuleAPIsCalled = false;
+			}
+		}
+	};
+};
+
+module.exports = {
+	create,
+	meta: {
```
Comment (sindresorhus) [inline rules/jsx-import.js]: You need to mark the type of the rule: https://github.com/sindresorhus/eslint-plugin-unicorn/commit/610a9014ab6a7076eaebf1990a143f15668ccba2

File: rules/jsx-import.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	while (scope.type !== 'global') {
+		scope = scope.upper;
+		variables = variables.concat(scope.variables);
+	}
+
+	if (scope.childScopes.length > 0) {
+		variables = scope.childScopes[0].variables.concat(variables);
+
+		if (scope.childScopes[0].childScopes.length > 0) {
+			variables = scope.childScopes[0].childScopes[0].variables.concat(variables);
```
Comment (futpib) [inline rules/jsx-import.js]: Would be nice to know if and why is this needed. Already seen it copy-pasted here https://github.com/sindresorhus/eslint-plugin-unicorn/pull/224#discussion_r251252397

Comment (sindresorhus) [reply rules/jsx-import.js]: Yeah, and would be good to have some separate tests for util methods.

**Comment by sindresorhus** [comment]
@doniyor2109 Ping

File: test/jsx-import.js (hunk 2 above)
Comment (doniyor2109) [reply test/jsx-import.js]: @sindresorhus If I add `dom-chef` to support React then when using below code won't be fixable.

`import React from 'foobar'`

Comment (sindresorhus) [reply test/jsx-import.js]: I think that's fine. Nobody would do `import React from 'foobar'` any way.

**Comment by sindresorhus** [comment]
@doniyor2109 There are still PRs comments that are not addressed and the code is not passing linting (run `npx xo --fix` locally).

**Comment by sindresorhus** [comment]
@doniyor2109 Do you want to finish this or should I close it?

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/eslint-plugin-unicorn/issues/155#issuecomment-559009476
